### ISSUE TITLE
Display PNG preview for widget

### ISSUE THREAD

**Issue by pkerpedjiev** [jupyter-widgets/ipywidgets#2682]
Hey! I have an IPython widget that somewhat resembles an interactive map. Whenever the user pans or zooms, the view changes. When the notebook is saved, stored and viewed on GitHub, I would like to see a preview of what was visible when the notebook was saved. 

AFAICT, to accomplish this I need to set `outputs["data"]["image/png"]` in the notebook's saved JSON representation. 

To do that I've tried the following:

1. On every widget change, it generates a screenshot of itself and stores that as a base64 string in a traitlet in the model.
2. This trait is `@observe`-ed and every time it changes I call `update_display`:

```
                self.display_data["image/png"] = change["new"]
                update_display(
                    self.display_data,
                    display_id=self.display_handle.display_id,
                    raw=True,
                )
```

This seems to work  except that it reloads the widget as well (i.e. it disappears and reappears again). My question is, is there a way to dynamically update the cell's data without reloading the widget?

**Comment by jasongrout** [comment]
I don't think so, since you are actually asking the system to rerender the displayed output with that update_display command.

**Comment by pkerpedjiev** [comment]
Hmm, is there no way to go around the `update_display` command and just change the cell's `display_data` directly?

Perhaps it could be snuck in on the client?

**Comment by jasongrout** [comment]
> Hmm, is there no way to go around the `update_display` command and just change the cell's `display_data` directly?

I think the answer to your question is no (as in I think I understand what you are trying to do in your workaround), but having a more comprehensive self-contained example to see *exactly* what you are doing would be helpful.

In particular, it seems that you are updating the actual display that contains a widget, so that the mimebundle has both the widget and a png?

> Perhaps it could be snuck in on the client?

That would probably be an issue for the particular client (notebook, lab, etc.), then.

**Comment by jasongrout** [comment]
That said, you can see previous discussion about this topic at https://github.com/jupyter-widgets/ipywidgets/issues/2280

**Comment by jasongrout** [comment]
(ah, from https://github.com/higlass/higlass-python/issues/49 it looks like you knew about #2280)

**Comment by pkerpedjiev** [comment]
> In particular, it seems that you are updating the actual display that contains a widget, so that the mimebundle has both the widget and a png?

Yes, that's exactly right. If I do this manually by going in changing the saved `.ipynb` file with a base64 png, I can load it just fine.

As you saw, there were a few threads I was looking at and what I wanted to accomplish seemed so tantalizingly close that I thought it best to just ask.

Anyway, thanks for your help. I'll poke around those threads some more. https://github.com/jupyter-widgets/ipywidgets/pull/314 looks promising.

**Comment by jasongrout** [comment]
Great, thanks for looking into this. A lot of people would love this. It seems like a tricky problem to solve in general, but perhaps an abstract method could be extended by each widget to provide a snapshot of itself, and that could be utilized in a general solution to save a snapshot of the top-level widget.
